runtime/vm/UpcallThunkMem.cpp

```diff
@@ -190,21 +225,46 @@ allocateThunkHeap(J9UpcallMetaData *data)
 		goto freeAllMemoryThenExit;
 	}
 
-	/* Store the heap handle in J9UpcallThunkHeapWrapper if the thunk heap is successfully created */
+	/* Store the heap handle in J9UpcallThunkHeapWrapper if the thunk heap is successfully created. */
 	thunkHeapWrapper->heap = thunkHeap;
 	thunkHeapWrapper->heapSize = pageSize;
 	thunkHeapWrapper->vmemID = vmemID;
-	vm->thunkHeapWrapper = thunkHeapWrapper;
+	thunkHeapNode->thunkHeapWrapper = thunkHeapWrapper;
+
+	/* Set the newly created thunk heap node as the list head if it doesn't exist
+	 * or when there is no free thunk heap to be allocated in the list for now.
+	 */
+	if (NULL == vm->thunkHeapHead) {
+		metaDataHashTable = hashTableNew(OMRPORT_FROM_J9PORT(vm->portLibrary), "Upcall metadata table", 0,
+				sizeof(J9UpcallMetaDataEntry), 0, 0, J9MEM_CATEGORY_VM_FFI, upcallMetaDataHashFn, upcallMetaDataEqualFn, NULL, NULL);
+		if (NULL == metaDataHashTable) {
+			Trc_VM_allocateThunkHeap_create_metadata_hash_table_failed();
+			goto freeAllMemoryThenExit;
+		}
+		thunkHeapNode->metaDataHashTable = metaDataHashTable;
+		vm->thunkHeapHead = thunkHeapNode;
+		vm->thunkHeapHead->next = vm->thunkHeapHead;
+	} else {
+		/* The hashtable is shared among all thunk heaps in the list. */
+		thunkHeapNode->metaDataHashTable = vm->thunkHeapHead->metaDataHashTable;
+		thunkHeapNode->next = vm->thunkHeapHead->next;
+		vm->thunkHeapHead->next = thunkHeapNode;
+		vm->thunkHeapHead = thunkHeapNode;
+	}
 
 #if defined(OSX) && defined(AARCH64)
 	pthread_jit_write_protect_np(1);
-#endif
+#endif /* defined(OSX) && defined(AARCH64) */
 
 done:
 	Trc_VM_allocateThunkHeap_Exit(thunkHeap);
-	return thunkHeap;
+	return thunkHeapNode;
 
 freeAllMemoryThenExit:
+	if (NULL != thunkHeapNode) {
```

Comment:
Freeing NULL is explicitly allowed in the port library, so no need for this test.

Reply:
Agreed and removed the test as suggested above.